Kernel/Devices/Input/HID/MouseDriver.cpp

```diff
@@ -0,0 +1,131 @@
+    InputManagement::the().attach_standalone_input_device(*handler->m_mouse_device);
+    return handler;
+}
+
+MouseDriver::MouseDriver(Device const& device, ::HID::ApplicationCollection const& application_collection, NonnullRefPtr<::MouseDevice> mouse_device)
+    : ApplicationCollectionDriver(device, application_collection)
+    , m_mouse_device(move(mouse_device))
+{
+}
+
+ErrorOr<void> MouseDriver::on_report(ReadonlyBytes report_data)
+{
+    MousePacket mouse_packet {};
+
+    TRY(::HID::parse_input_report(m_device.report_descriptor(), m_application_collection, report_data, [&mouse_packet](::HID::Field const& field, i64 value) -> ErrorOr<IterationDecision> {
+        u32 usage = 0;
+        if (field.is_array) {
+            if (!field.usage_minimum.has_value())
+                return Error::from_errno(ENOTSUP); // TODO: What are we supposed to do here?
+
+            usage = value + field.usage_minimum.value();
+            value = 1;
+        } else {
+            usage = field.usage.value();
+        }
+
+        using enum Usage;
+        switch (static_cast<Usage>(usage)) {
+        // FIXME: Do we need to handle relative button items? That would mean the mouse sends us On/Off toggle events for its buttons.
+        //        Mice should usually have the Absolute and Preferred State flags set.
+        case Button1:
+            if (value == 1)
+                mouse_packet.buttons |= MousePacket::Button::LeftButton;
+            else
+                mouse_packet.buttons &= ~MousePacket::Button::LeftButton;
+            break;
+
+        case Button2:
+            if (value == 1)
+                mouse_packet.buttons |= MousePacket::Button::RightButton;
+            else
+                mouse_packet.buttons &= ~MousePacket::Button::RightButton;
+            break;
+
+        case Button3:
+            if (value == 1)
+                mouse_packet.buttons |= MousePacket::Button::MiddleButton;
+            else
+                mouse_packet.buttons &= ~MousePacket::Button::MiddleButton;
+            break;
+
+        case Button4:
+            if (value == 1)
+                mouse_packet.buttons |= MousePacket::Button::BackwardButton;
+            else
+                mouse_packet.buttons &= ~MousePacket::Button::BackwardButton;
+            break;
+
+        case Button5:
+            if (value == 1)
+                mouse_packet.buttons |= MousePacket::Button::ForwardButton;
+            else
+                mouse_packet.buttons &= ~MousePacket::Button::ForwardButton;
+            break;
+
+        case X:
+            mouse_packet.is_relative = field.is_relative;
+            if (field.is_relative)
+                mouse_packet.x = value;
+            else
+                mouse_packet.x = static_cast<int>((value - field.logical_minimum) * 0xffff / (field.logical_maximum - field.logical_minimum));
+
```

Comment:
I have the feeling the physical range should matter here, but Ig thats only the case if you care for and support the advertised unit

Reply:
Yeah, we only use the logical values currently.